### PR TITLE
LibSQL: Implement a table description statement

### PR DESCRIPTION
**Pull request by i3abghany**
This adds the support for a non-standard `DESCRIBE TABLE TableName;` statement in our DBMS implementation. Even though it's non-standard, practically every DBMS out there has a sort of table-description user-facing mechanism. This is done by simply having an internal schema that defines the relation having information about a table. For now, it only contains column names and their data types.

Screenshot from 2021-10-05 21-41-17

### CODE REVIEW

**Comment by i3abghany** [comment]
@JanDeVisser Hello, can you please help me with this issue in the CI when you have some free time? I don't get why exactly do I trigger undefined behaviour in `IntegerImpl::compare`? I followed the stack trace but I don't seem to find the problem

**Comment by i3abghany** [comment]
@trflynn89 any help is appreciated if you have some time! <3

**Comment by trflynn89** [comment]
Since your branch is 1283 commits behind master, the first thing I would do is rebase on master and see if the issue still occurs

**Comment by i3abghany** [comment]
> Since your branch is 1283 commits behind master, the first thing I would do is rebase on master and see if the issue still occurs

I've done so but unfortunately, the problem still occurs

**Comment by JanDeVisser** [comment]
I’m not surprised funky stuff happens there. I honestly tried to be careful
but there’s some jiggery-pokery going on there. Unfortunately I’m a 60km
bike ride away from my machine right now; will get back in touch once I
have the code in front of me

**Comment by trflynn89** [comment]
I'm thinking this isn't an issue with your code, but with the underlying `IntegerImpl::compare` method. If `value1` has the integer `-232937294`, and `value2` has the integer `2068933544` (which is totally valid), then `value1 - value2` will certainly overflow.

We can change `IntegerImpl::compare` to avoid signed arithmetic:
```c++
int IntegerImpl::compare(Value const& other) const
{
    auto casted = other.to_int();
    if (!casted.has_value())
        return 1;

    if (value() == casted.value())
        return 0;
    return value() < casted.value() ? -1 : 1;
}
```

**Comment by i3abghany** [comment]
> I'm thinking this isn't an issue with your code, but with the underlying `IntegerImpl::compare` method. If `value1` has the integer `-232937294`, and `value2` has the integer `2068933544` (which is totally valid), then `value1 - value2` will certainly overflow.
> 
> We can change `IntegerImpl::compare` to avoid signed arithmetic:
> 
> ```c++
> int IntegerImpl::compare(Value const& other) const
> {
>     auto casted = other.to_int();
>     if (!casted.has_value())
>         return 1;
> 
>     if (value() == casted.value())
>         return 0;
>     return value() < casted.value() ? -1 : 1;
> }
> ```

That's a more suitable solution I believe and I can implement it in my PR if you want as a quick fix for the issue. However, how do we even get there? I didn't touch the test that fails, so we should expect that the test should fail with or without the PR, right? So even if `compare` could be improved there might still be a problem elsewhere that triggers this subtraction in the first place.

**Comment by trflynn89** [comment]
The issue presents here due the `internal_describe_table` table being added. So when the test creates its own table, `CreateTable::execute` first calls `Database::get_table` to see if it already exists. This lookup involves comparing the internal keys of the `internal_describe_table` table and the new table being created, which ends up in `IntegerImpl::compare`.

Also note that (if I'm reading this correctly) the integer values being created for the internal keys are hashes of the schema/table names. Looking at the failure [here](https://dev.azure.com/SerenityOS/SerenityOS/_build/results?buildId=4546&view=logs&jobId=97f87b91-8e79-5621-6312-f3af39cfde0b&j=2566f013-1391-5cbe-90ce-ae4ec398ae06&t=dc6c2af9-d505-5d29-ab30-532bffaa89e3) the error is:
```
runtime error: signed integer overflow: -232937294 - 2068933544 cannot be represented in type 'int'
```
The value `-232937294` is the hash of the schema name `"master"` (cast to `int`), so that's why we're hitting it now.

So in other words we were just lucky to have not hit this before :)

> I can implement it in my PR if you want as a quick fix for the issue

Yep, go ahead! You can add that as a separate commit in this PR.

**Comment by JanDeVisser** [comment]
BTW I take full responsibility for being sloppy with signed and unsigned ints. I will do the required self flagellation shortly and promise being more careful in the future. 

On a more serious note: I think we need an Unsigned type to avoid things like this in the future.

**Comment by i3abghany** [comment]
> The issue presents here due the `internal_describe_table` table being added. So when the test creates its own table, `CreateTable::execute` first calls `Database::get_table` to see if it already exists. This lookup involves comparing the internal keys of the `internal_describe_table` table and the new table being created, which ends up in `IntegerImpl::compare`.
> 
> Also note that (if I'm reading this correctly) the integer values being created for the internal keys are hashes of the schema/table names. Looking at the failure [here](https://dev.azure.com/SerenityOS/SerenityOS/_build/results?buildId=4546&view=logs&jobId=97f87b91-8e79-5621-6312-f3af39cfde0b&j=2566f013-1391-5cbe-90ce-ae4ec398ae06&t=dc6c2af9-d505-5d29-ab30-532bffaa89e3) the error is:
> 
> ```
> runtime error: signed integer overflow: -232937294 - 2068933544 cannot be represented in type 'int'
> ```
> 
> The value `-232937294` is the hash of the schema name `"master"`, and `2068933544` is the hash of the schema name `"TESTSCHEMA"` used by the test (each cast to `int`), so that's why we're hitting it now.
> 
> So in other words we were just lucky to have not hit this before :)
> 
> > I can implement it in my PR if you want as a quick fix for the issue
> 
> Yep, go ahead! You can add that as a separate commit in this PR.

Thank you so much for the clarification!

File: Tests/LibSQL/TestNonSqlStatementParsing.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
+/*
```
Comment (trflynn89) [inline Tests/LibSQL/TestNonSqlStatementParsing.cpp:1]: I don't think this needs to be in its own file, it can just got in TestSqlStatementParser.cpp. It's still a SQL statement even if it's our own thing :)

Would also be nice to have a test case added to TestSqlStatementExecution.cpp to validate the execution of this statement.

**Comment by i3abghany** [comment]
I'll certainly address the feedback... I just couldn't allocate enough time in the past month, and I'll have some free time soon.

**Review by trflynn89 — APPROVED** [review]

**Comment by trflynn89** [comment]
@i3abghany If you didn't notice (I don't think GitHub sends notifications for this), your PR has merge conflicts with `master`, so you'll have to rebase before this can be merged

**Comment by i3abghany** [comment]
> @i3abghany If you didn't notice (I don't think GitHub sends notifications for this), your PR has merge conflicts with `master`, so you'll have to rebase before this can be merged

 Thanks for the reminder, I didn't realize that it was in conflict. I hopefully fixed them and did very minor modifications accordingly.

File: Userland/Libraries/LibSQL/Database.cpp [hunk 2]
```diff
@@ -47,6 +48,21 @@ ErrorOr<void> Database::open()
...
+        TRY(add_schema(*master_schema));
+    }
+
+    auto describe_internal_table = TableDef::construct(master_schema, "internal_describe_table");
+    describe_internal_table->append_column("Name", SQLType::Text);
+    describe_internal_table->append_column("Type", SQLType::Text);
+
+    // `add_table` errors only on recreating an already-existing table,
+    //  which we don't care about.
+    auto discarded_return_value = add_table(*describe_internal_table);
```
Comment (trflynn89) [inline Userland/Libraries/LibSQL/Database.cpp:64]: I'm kinda surprised this doesn't also raise a warning about an unused variable. But I'm not sure we want to suppress errors here anyways?
```c++
TRY(add_table(*describe_internal_table));

Comment (i3abghany) [reply Userland/Libraries/LibSQL/Database.cpp:64]: For some reason, the table gets created repeatedly (I don't exactly get why)... So using `TRY` caused problems as it errs and logs "duplicate table" or something similar.

Should I check if the table exists, and if not so create it? I manually checked `is_error` on the return value of `add_table` but it's not so meaningful as we want to ignore the error (`add_table` probably only errs on a recreation of an already existing table). So what do you suggest?

**Comment by i3abghany** [comment]
(sorry if this gets mangled... I'm replying through email)For some reason that I don't understand, the "internal_describe_table" gets created multiple times, which raises an error so if I use `TRY` it breaks. Btw I didn't get any warnings as far as I can tell.Should I check if the table exists and only create it if not so?On 10 Dec 2021 21:50, Tim Flynn ***@***.***> wrote:
@trflynn89 commented on this pull request.



In Userland/Libraries/LibSQL/Database.cpp:
> @@ -47,6 +48,21 @@ ErrorOr<void> Database::open()
         default_schema = SchemaDef::construct("default");
         TRY(add_schema(*default_schema));
     }
+
+    auto master_schema = TRY(get_schema("master"));
+    if (!master_schema) {
+        master_schema = SchemaDef::construct("master");
+        TRY(add_schema(*master_schema));
+    }
+
+    auto describe_internal_table = TableDef::construct(master_schema, "internal_describe_table");
+    describe_internal_table->append_column("Name", SQLType::Text);
+    describe_internal_table->append_column("Type", SQLType::Text);
+
+    // `add_table` errors only on recreating an already-existing table,
+    //  which we don't care about.
+    auto discarded_return_value = add_table(*describe_internal_table);

I'm kinda surprised this doesn't also raise a warning about an unused variable. But I'm not sure we want to suppress errors here anyways?
TRY(add_table(*describe_internal_table));

File: Userland/Libraries/LibSQL/Database.cpp (hunk 2 above)
Comment (trflynn89) [reply Userland/Libraries/LibSQL/Database.cpp:64]: I'd assume it already exists if the database file on disk already exists.

So on a brand new database, we successfully create the internal DESCRIBE table, then at some point later we persist that database to disk. Then when we "create" the database again from that file, the table already exists, and we get the error.

I'd suggest checking if it exists before adding it, that way we can propagate any errors creating the table up to the caller. There might not be any other errors now, but that could change in the future, and I'd rather not mask real errors.

just my opinion, @JanDeVisser can chime in if he has another opinion :)

**Comment by awesomekling** [comment]
Hmm, what's the state of this PR?

**Comment by i3abghany** [comment]
@awesomekling It should be ready once I know what to do with the error here. I was waiting for @trflynn89 's suggestion on what to do there - I spent some time trying to see what's causing the table to be created multiple times and just can't wrap my head around it - and I think that sort of "catching" the _duplicate table error_ and not doing anything about it is good enough? No other errors can occur as far as I can tell.

So to sum up: It's ready, but I was waiting to see what to do about handling this error.

**Comment by trflynn89** [comment]
@i3abghany I replied a few weeks ago: https://github.com/SerenityOS/serenity/pull/10371#discussion_r768020848

**Comment by i3abghany** [comment]
Great - I'm going through final exams now, and I'm going to address it when I finish. Should I close this PR for now and reopen it when it's done?

**Comment by trflynn89** [comment]
Nah you can leave it open - good luck on exams!

**Comment by i3abghany** [comment]
Great, thanks!

**Comment by i3abghany** [comment]
Azure jobs timed out and crashed after 1 hour, what can I do in this case?

File: Userland/Libraries/LibSQL/Database.cpp [hunk 3]
```diff
@@ -47,6 +48,24 @@ ErrorOr<void> Database::open()
...
+    auto master_schema = TRY(get_schema("master"));
+    if (!master_schema) {
+        master_schema = SchemaDef::construct("master");
+        TRY(add_schema(*master_schema));
+    }
+
+    auto table_def_or_error = get_table("master", "internal_describe_table");
+    if (table_def_or_error.is_error())
+        return table_def_or_error.release_error();
+    auto table_def = table_def_or_error.release_value();
```
Comment (trflynn89) [inline Userland/Libraries/LibSQL/Database.cpp:61]: This method itself returns `ErrorOr<>`, so couldn't these lines here  just be:
```c++
auto table_def = TRY(get_table("master", "internal_describe_table"));
```

**Review by trflynn89 — APPROVED** [review]

**Comment by i3abghany** [comment]
no changes - rebasing to rerun ci

**Comment by trflynn89** [comment]
> no changes - rebasing to rerun ci

I think the tests are legitimately broken on `master` right now

**Comment by linusg** [comment]
This was fixed on master, please rebase the PR again so we can get a green CI run :)

**Comment by i3abghany** [comment]
@linusg Done!

**Comment by i3abghany** [comment]
@linusg addressed merge conflicts.
